File: DataStructures/src/main/java/com/zhangdd/sort/InsertSort.java
package com.zhangdd.sort;

/**
 * @author zhangdd on 2021/10/30
 */
public class InsertSort {

    public static void main(String[] args) {
        //从小到大排
        int[] arr = {1, 9, 7, -3, 2};

        //从j=1开始
        for (int j = 1; j < arr.length; j++) {
            int readyIndex = j;
            int readyValue = arr[j];//暂存开始时索引的值，下面的循环是为其寻找合适位置
            for (int i = j - 1; i >= 0; i--) {
                if (arr[i] > readyValue) {
                    arr[i + 1] = arr[i];
                    readyIndex--;
                } else {
                    break;
                }
            }
            //找到合适的下标位置，设置值
            arr[readyIndex] = readyValue;
        }

        for (int j : arr) {
            System.out.printf("%d ", j);
        }
    }
}

File: python/Algorithm/sort/InsertSort.py
#!/usr/bin/env python3
"""
@author: zhangdd on 2022/2/3
"""


def insert_sort(arr):
    for i in range(1, len(arr)):
        ready_index = i
        ready_value = arr[i]
        for j in range(i - 1, -1, -1):
            if arr[j] > ready_value:
                arr[j + 1] = arr[j]
                ready_index -= 1
            else:
                break
        arr[ready_index] = ready_value


if __name__ == '__main__':
    collection = [1, 9, 7, -3, 2]
    insert_sort(collection)
    print(collection)
